### PR TITLE
Issue/4802 add method and or arguments to compilerservice to export with specified exporter plugin

### LINKED ISSUE #4802: add method/arguments to compilerservice to export with specified exporter plugin
Make sure to allow it for both exporting and non-exporting compiles.

### PR DESCRIPTION
**Pull request by Hugo-Inmanta**
# Description

Add argument to compilerservice to allow exporting with specified exporter plugin.

closes #4802
# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
~~- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )~~

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

### CODE REVIEW

File: changelogs/unreleased/4802-add-method-and-or-arguments-to-compilerservice-to-export-with-specified-exporter-plugin.yml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+description: Add argument to compilerservice to allow exporting with the specified exporter plugin
+change-type: patch
```
Comment (arnaudsjs) [inline changelogs/unreleased/4802-add-method-and-or-arguments-to-compilerservice-to-export-with-specified-exporter-plugin.yml:3]:
```suggestion
change-type: minor
```

File: src/inmanta/db/versions/v202209160.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+"""
+
+from asyncpg import Connection
+
+
+async def update(connection: Connection) -> None:
+    schema = """
+
+    ALTER TABLE public.compile
+        ADD COLUMN exporter_plugin varchar DEFAULT '';
```
Comment (arnaudsjs) [inline src/inmanta/db/versions/v202209160.py:26]: This is inconsistent with the default value in the ORM.
```suggestion
        ADD COLUMN exporter_plugin varchar DEFAULT NULL;
```

File: src/inmanta/db/versions/v202209160.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    """
+    async with connection.transaction():
+        await connection.execute(schema)
```
Comment (arnaudsjs) [inline src/inmanta/db/versions/v202209160.py:29]: The database migration tool already starts a transaction. There is no need to start an extra subtransaction.
```suggestion
    await connection.execute(schema)
```

File: tests/db/test_v202208190_to_v202209160.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+    Contact: code@inmanta.com
+"""
+import os
+from typing import Awaitable, Callable, List
+
+import pytest
+
+
+@pytest.mark.db_restore_dump(os.path.join(os.path.dirname(__file__), "dumps/v202208180.sql"))
```
Comment (arnaudsjs) [inline tests/db/test_v202208190_to_v202209160.py:24]:
```suggestion
@pytest.mark.db_restore_dump(os.path.join(os.path.dirname(__file__), "dumps", "v202208180.sql"))
```

File: tests/db/test_v202208190_to_v202209160.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+async def test_added_exporter_plugin(
+    migrate_db_from: Callable[[], Awaitable[None]],
+    get_columns_in_db_table: Callable[[str], Awaitable[List[str]]],
+    get_custom_postgresql_types: Callable[[], Awaitable[List[str]]],
+) -> None:
```
Comment (arnaudsjs) [inline tests/db/test_v202208190_to_v202209160.py:29]: This fixture is not used by this test case.
```suggestion
    get_columns_in_db_table: Callable[[str], Awaitable[List[str]]],
) -> None:
```

File: tests/server/test_compilerservice.py [hunk 6]
```diff
@@ -119,6 +120,54 @@ def write_main(self, main: str, environment: Optional[data.Environment] = None)
...
+            new_content_init_py=plugin_code,
+            new_name=plugin_name,
+        )
+        subprocess.check_output(["git", "add", f"{self.libs_dir}"], cwd=self.src_dir)
+        subprocess.check_output(["git", "commit", "-m", "write plugin", "--allow-empty"], cwd=self.src_dir)
+
+
+async def compile_and_assert(
+    env, client, project_work_dir: str, export=True, meta={}, env_vars={}, update=False, exporter_plugin=None
+):
```
Comment (arnaudsjs) [inline tests/server/test_compilerservice.py:136]: This method can use a docstring.

File: tests/server/test_compilerservice.py [hunk 7]
```diff
@@ -119,6 +120,54 @@ def write_main(self, main: str, environment: Optional[data.Environment] = None)
...
 
+    def write_plugin(self, plugin_name: str, plugin_code: str, modules_dir):
+        v1_module_from_template(
+            source_dir=modules_dir,
+            dest_dir=os.path.join(self.libs_dir, plugin_name),
+            new_content_init_py=plugin_code,
+            new_name=plugin_name,
+        )
+        subprocess.check_output(["git", "add", f"{self.libs_dir}"], cwd=self.src_dir)
+        subprocess.check_output(["git", "commit", "-m", "write plugin", "--allow-empty"], cwd=self.src_dir)
```
Comment (arnaudsjs) [inline tests/server/test_compilerservice.py:131]:
```suggestion
    def add_v1_module(self, plugin_name: str, plugin_code: str, modules_dir: str) -> None:
        v1_module_from_template(
            source_dir=modules_dir,
            dest_dir=os.path.join(self.libs_dir, plugin_name),
            new_content_init_py=plugin_code,
            new_name=plugin_name,
        )
        subprocess.check_output(["git", "add", f"{self.libs_dir}"], cwd=self.src_dir)
        subprocess.check_output(["git", "commit", "-m", "add_v1_module", "--allow-empty"], cwd=self.src_dir)
```

File: src/inmanta/db/versions/v202209160.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+    Contact: code@inmanta.com
+"""
+
+from asyncpg import Connection
+
+
+async def update(connection: Connection) -> None:
+    schema = """
+
```
Comment (sanderr) [inline src/inmanta/db/versions/v202209160.py:24]:
```suggestion
```

File: src/inmanta/db/versions/v202209160.py (hunk 2 above)
Comment (sanderr) [inline src/inmanta/db/versions/v202209160.py:26]:
```suggestion
        ADD COLUMN exporter_plugin varchar DEFAULT NULL;
```

File: tests/db/dumps/v202209160.sql [hunk 9]
```diff
@@ -0,0 +1,1111 @@
...
+    'success',
+    'failed',
+    'deploying',
+    'pending'
+);
+
+
+SET default_tablespace = '';
+
+SET default_table_access_method = heap;
```
Comment (sanderr) [inline tests/db/dumps/v202209160.sql:109]:
```suggestion
--SET default_table_access_method = heap;
```
See https://github.com/inmanta/inmanta-core/blob/19bac9ed9c0790522a7ef0c985c464d90c74db1a/tests/db/dump_tool.py#L44

File: src/inmanta/export.py [hunk 10]
```diff
@@ -417,7 +417,7 @@ def add_resource(self, resource: Resource) -> None:
         """
         Add a new resource to the list of exported resources. When
         commit_resources is called, the entire list of resources is send
```
Comment (sanderr) [inline src/inmanta/export.py:419]:
```suggestion
        commit_resources is called, the entire list of resources is sent
```

File: tests/db/test_v202208190_to_v202209160.py [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Contact: code@inmanta.com
+"""
+import os
+from typing import Awaitable, Callable, List
```
Comment (sanderr) [inline tests/db/test_v202208190_to_v202209160.py:19]: We concluded that for new code we would use `from collections import abc` and then use `abc.Awaitable`, `abc.Callable`, `abc.Sequence`, ... where applicable and `list`, `dict` where required.

File: tests/db/test_v202208190_to_v202209160.py (hunk 4 above)
Comment (sanderr) [inline tests/db/test_v202208190_to_v202209160.py:24]:
```suggestion
@pytest.mark.db_restore_dump(os.path.join(os.path.dirname(__file__), "dumps", "v202208180.sql"))
```

File: tests/server/test_compilerservice.py [hunk 12]
```diff
@@ -119,6 +120,54 @@ def write_main(self, main: str, environment: Optional[data.Environment] = None)
         subprocess.check_output(["git", "add", "main.cf"], cwd=self.src_dir)
         subprocess.check_output(["git", "commit", "-m", "write main.cf", "--allow-empty"], cwd=self.src_dir)
 
+    def write_plugin(self, plugin_name: str, plugin_code: str, modules_dir):
```
Comment (sanderr) [inline tests/server/test_compilerservice.py:123]: Perhaps we can make this a bit more generic so we can more easily extend it in the future if we need it?
```suggestion
    def write_module(self, module_name: str, *, plugin_code: str, template_dir: str) -> None:
```

File: tests/server/test_compilerservice.py (hunk 6 above)
Comment (sanderr) [inline tests/server/test_compilerservice.py:136]: This could use a docstring.

Comment (sanderr) [inline tests/server/test_compilerservice.py:136]:
```suggestion
) -> tuple[CompileRun, abc.Mapping[str, object]]:
```

**Review by sanderr — CHANGES_REQUESTED** [review]
Some small remarks, mostly looks good to me.

File: tests/server/test_compilerservice.py [hunk 13]
```diff
@@ -1285,3 +1312,59 @@ def patch_get_protected_inmanta_packages():
...
+
+    environment_factory.write_plugin(used_plugin_name, make_plugin_code(used_exporter), module_template)
+    environment_factory.write_plugin(unused_plugin_name, make_plugin_code(unused_exporter), module_template)
+
+    project_work_dir = os.path.join(tmpdir, "work")
+    ensure_directory_exist(project_work_dir)
+
+    # with export
+    compile, stages = await compile_and_assert(
+        env=env, project_work_dir=project_work_dir, client=client, export=True, exporter_plugin="test_exporter"
```
Comment (arnaudsjs) [inline tests/server/test_compilerservice.py:1356]:
```suggestion
        env=env, project_work_dir=project_work_dir, client=client, export=True, exporter_plugin=used_exporter
```

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: tests/server/test_compilerservice.py [hunk 14]
```diff
@@ -1285,3 +1312,59 @@ def patch_get_protected_inmanta_packages():
...
+    used_exporter = "test_exporter"
+    unused_exporter = "unused_test_exporter"
+
+    used_plugin_name = used_exporter + "_plugin"
+    unused_plugin_name = unused_exporter + "_plugin"
+
+    def make_main():
+        return f"""
+import {used_plugin_name}
+        """
```
[the remark this answers is not in the document]
Comment (Hugo-Inmanta) [reply tests/server/test_compilerservice.py:1333]: Aah right

File: tests/db/test_v202208190_to_v202209160.py [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+
+    Contact: code@inmanta.com
+"""
+import os
+from collections import abc
+
+import pytest
+
+
+@pytest.mark.db_restore_dump(os.path.join(os.path.dirname(__file__), "dumps", "v202208180.sql"))
```
Comment (arnaudsjs) [inline tests/db/test_v202208190_to_v202209160.py:24]: In #4829 the latest version of the database schema changed to v202209090. I think we need to test this change from that version and generate a new database dump.

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: tests/db/test_v202208190_to_v202209160.py (hunk 15 above)
Comment (arnaudsjs) [reply tests/db/test_v202208190_to_v202209160.py:24]: Hmm. The database dump created in #4829 seems to have an invalid version in its name. Do you fix that?

Comment (Hugo-Inmanta) [reply tests/db/test_v202208190_to_v202209160.py]: Done!

**Review by arnaudsjs — APPROVED** [review]

**Pull request closed without merging**